cmd/core/pfcp_hearbeat.go

```diff
@@ -0,0 +1,54 @@
+package core
+
+import (
+	"github.com/wmnsk/go-pfcp/ie"
+	"github.com/wmnsk/go-pfcp/message"
+	"log"
+	"net"
+	"time"
+)
+
+func HandlePfcpHeartbeatRequest(conn *PfcpConnection, msg message.Message, addr string) (message.Message, error) {
+	hbreq := msg.(*message.HeartbeatRequest)
+	if association := conn.GetAssociation(addr); association != nil {
+		association.CheckInContact()
+	}
+	ts, err := hbreq.RecoveryTimeStamp.RecoveryTimeStamp()
+	if err != nil {
+		log.Printf("Got Heartbeat Request with invalid TS: %s, from: %s", err, addr)
+		return nil, err
+	} else {
+		log.Printf("Got Heartbeat Request with TS: %s, from: %s", ts, addr)
+	}
+
+	hbres := message.NewHeartbeatResponse(hbreq.SequenceNumber, ie.NewRecoveryTimeStamp(time.Now()))
+	log.Printf("Sent Heartbeat Response to: %s", addr)
+	return hbres, nil
+}
+
+func HandlePfcpHeartbeatResponse(conn *PfcpConnection, msg message.Message, addr string) (message.Message, error) {
+	hbresp := msg.(*message.HeartbeatResponse)
+	ts, err := hbresp.RecoveryTimeStamp.RecoveryTimeStamp()
+	if err != nil {
+		log.Printf("Got Heartbeat Response with invalid TS: %s, from: %s", err, addr)
+	} else {
+		log.Printf("Got Heartbeat Response with TS: %s, from: %s", ts, addr)
+	}
+	if association := conn.GetAssociation(addr); association != nil {
```

Comment:
Why in response handler GetAssociation is called after recovery timestamp processing?

Reply:
I've forgotten a return if timestamp is corrupted.